gits/terminal.py

```diff
@@ -213,46 +252,66 @@ def _default_rendition(self):
         """
         self._cap_set_color(0)
 
+    def _set_bg_color(self, color):
+        """Sets background color. """
+        color_bits, _ = divmod(self._sgr, MAGIC_NUMBER)
+        _, fg = divmod(color_bits, 16)
+        new_color_bits = color * 16 + fg
+        self._sgr &= ~(color_bits << 40)  # clear color bits
+        self._sgr |= new_color_bits << 40  # update bg and fg colors
+
+    def _set_fg_color(self, color):
+        """Sets foreground color. """
+        color_bits, _ = divmod(self._sgr, MAGIC_NUMBER)
+        bg, _ = divmod(color_bits, 16)
+
+        # Switch to the bright color scheme if necessary.
+        if not self._normal_mode:
+            color += 8
+
+        new_color_bits = bg * 16 + color
+        self._sgr &= ~(color_bits << 40)  # clear color bits
+        self._sgr |= new_color_bits << 40  # update bg and fg colors
+
     def _cap_set_color_pair(self, p1, p2):
         if p1 == 0 and p2 == 10:  # sgr0
-            self._sgr = MAGIC_NUMBER
+            self._sgr = BLACK_AND_WHITE
         elif p1 == 39 and p2 == 49:  # op
-            self._sgr = MAGIC_NUMBER
+            self._sgr = BLACK_AND_WHITE
         else:
             self._cap_set_color(p1)
             self._cap_set_color(p2)
 
-    def _cap_set_color(self, colour):
-        if colour == 0:
-            self._sgr = MAGIC_NUMBER
-        elif colour == 1:  # bold
-            self._sgr = (self._sgr | 0x08000000)
-        elif colour == 2:  # dim
-            pass
-        elif colour == 4:  # smul
+    def _cap_set_color(self, color):
+        if color == 0:
+            self._sgr = BLACK_AND_WHITE
+        elif color == 1:  # bold
+            self._set_bit(BOLD_BIT)
+            self._cap_set_color(37)  # bold also means extra bright
+        elif color == 2:  # dim
             pass
-        elif colour == 5:  # blink
+        elif color == 4:  # smul, start underscore mode
+            self._set_bit(UNDERLINE_BIT)
```

Comment:
Different sources related to terminals use `underline` or `underscore`. For example, `man infocmp` uses `underline` and [1] uses both. I suggest we rename `underscore` into `underline` because the latter is well-spread in our code.

[1] http://pubs.opengroup.org/onlinepubs/7908799/xcurses/terminfo.html